### ISSUE TITLE
Changing VTX settings from script causes issues with VTX Table

### ISSUE THREAD

**Issue by AndreaPro** [betaflight/betaflight-tx-lua-scripts#486]
I am using Betaflight Lua Scripts 1.7.0 (I've even tried the nightly), Betaflight Configurator 10.9.0, Radiomaster TX16S with EdgeTX 2.8.5 and a BetaFPV Meteor 75 Pro updated to BF 4.4.2 and I am using ELRS 3.3.0 on BetaFPV Micro 1W TX and on the quad.
If I open the lua script and I adjust some settings in the VTX section and save them and I power off the quad and connect it to Betaflight configurator the VTX table is wiped completely and the settings are not stored. I've tried generating the lua script VTX table from the configurator and retrieving them from the quad but the result is the same. Pratically the VTX section of the script is unusable because after restart the VTX table is completely wiped.

**Comment by haslinghuis** [comment]
Duplicate of https://github.com/betaflight/betaflight-tx-lua-scripts/issues/482

**Comment by AndreaPro** [comment]
> Duplicate of #482

In my opinion is not a duplicate, it is a different issue. I have compiled the firmware with VTX and OSD already

**Comment by klutvott123** [comment]
Does this happen when making the same changes in the betaflight configurator?

**Comment by AndreaPro** [comment]
No, it happens only when making changes to VTX parameters using the lua
script from the radio. I've tried changing accelerometer trims and PID
profiles without issues

**Comment by klutvott123** [comment]
can you try downloading the vtx tables again? https://github.com/betaflight/betaflight-tx-lua-scripts#setting-up-vtx-tables

**Comment by AndreaPro** [comment]
Yes, I've tried multiple times both downloading them from the script on the
radio and even generating the lua script from the configurator and copied
it manually on the radio but the outcome is the same, as soon as I do one
change the VTX table is altered and becomes unusable

**Comment by klutvott123** [comment]
That's strange. I'll have to see if I can reproduce

**Comment by AndreaPro** [comment]
If there's something that I can do to troubleshoot the issue just let me
know

**Comment by klutvott123** [comment]
Is it the setting of a specific parameter that causes the issue? Could you try 4.4.1 and maybe 4.4.0?

**Comment by AndreaPro** [comment]
It happened when changing VTX power. I can try to go back to these versions
and let you know

**Comment by AndreaPro** [comment]
I have downgraded to BF 4.4.0 that was the previous release available for
my board, and it happens the same. I've noticed one thing, that if I copy
the VTX table lua script generated by Betaflight inside the script it works
until I change a parameter, but if I download the VTX table from the quad
it downloads a different VTX table that is the wiped one that I found
inside the configurator. Attached there is the Lua script generated by
Betaflight configurator and the lua script that is created inside the radio
if I download the VTX table. And there is the VTX table json saved before
and after the issue.

Il giorno ven 1 set 2023 alle ore 23:52 Andrea Proietti <
***@***.***> ha scritto:

> It happened when changing VTX power. I can try to go back to these
> versions and let you know
>
> On Fri, 1 Sept 2023, 23:48 Hans Christian Olaussen, <

**Comment by klutvott123** [comment]
Seems like the lua scripts end up with the "empty" vtx table after downloading? We'll have to figure out why that happens then.
I just tested this with a betafpv meteor65 board with 4.4.2, and I'm not able to reproduce. Vtx table is correctly downloaded, and I can change all parameters.

**Comment by klutvott123** [comment]
@AndreaPro Can you share a diff please?

**Comment by AndreaPro** [comment]
> @AndreaPro Can you share a diff please?

For sure! I have tried both with the stock CLI provided by BetaFPV and with my custom settings but I have the same issue.

[CLI.zip](https://github.com/betaflight/betaflight-tx-lua-scripts/files/12504952/CLI.zip)

**Comment by tomekste5** [comment]
Got the same issue if I download the vtx table or copy the lua script from betaflight to my controller, it overwrites my vtx table.

**Comment by AndreaPro** [comment]
> Got the same issue if I download the vtx table or copy the lua script from betaflight to my controller, it overwrites my vtx table.

Could you please share more infos about radio firmware version, Betaflight version, script version, drone model, etc...?

**Comment by tomekste5** [comment]
Oh ye sure sry,
Betaflight version: 4.3
Edge TX version: 2.8.0 (Radiomaster Boxer)
Betaflight lua script version: 1.7
Flight controller: Speedybee F405
VTX: TBS Unify Pro32 (https://www.team-blacksheep.com/vtx-table-for-betaflight.txt)

After I try to switch a setting, my VTX table on the flight controller gets overwritten to the same table u shared

**Comment by klutvott123** [comment]
The only thing I can think of is that the payload somehow contains more bytes than it should, and this happens:
https://github.com/betaflight/betaflight/blob/99be07729963e4b5e5cb540555450bda28bd2781/src/main/msp/msp.c#L3380-L3407

It's strange because in the BF lua scripts we're only sending four bytes back to the FC

**Comment by klutvott123** [comment]
@tomekste5 Which RX system are you using?

**Comment by tomekste5** [comment]
I am using Express LRS version 3.3 2.4GHz and as a Receiver I use the Radiomaster RP1

**Comment by AndreaPro** [comment]
> I am using Express LRS version 3.3 2.4GHz and as a Receiver I use the Radiomaster RP1

So we have in common ELRS 3.30 2.4GHz

**Comment by klutvott123** [comment]
maybe try disabling vtx administrator in the elrs lua script.

**Comment by AndreaPro** [comment]
> maybe try disabling vtx administrator in the elrs lua script.

I have tried VTX administrator in elrs lua script because the Betaflight script was destroying the VTX table, but I have never used it before discovering this issue

**Comment by klutvott123** [comment]
try disabling it by setting "band" to "off". did this happen with earlier versions of elrs? the payload sent back from the betaflight lua scripts is only 4 bytes, but if more bytes are sent the vtx tables could be reset. not sure where those bytes are coming from. needs more investigating.

**Comment by AndreaPro** [comment]
> try disabling it by setting "band" to "off". did this happen with earlier versions of elrs? the payload sent back from the betaflight lua scripts is only 4 bytes, but if more bytes are sent the vtx tables could be reset. not sure where those bytes are coming from. needs more investigating.

I have tried the Betaflight lua directly starting with ELRS 3.3.0. I will try disabling it but I am sure that the first time that I have noticed this issue I haven't enabled VTX admin yet

**Comment by tomekste5** [comment]
> maybe try disabling vtx administrator in the elrs lua script.
I have never enabled VTX administrator

**Comment by klutvott123** [comment]
If it's not too much hassle, could you guys please try ELRS 3.0? I'm not saying elrs is the issue, but we need to rule things out. As I said, the payload sent from the bf lua scripts is 4 bytes, but it could be that zero padding or something is added somewhere along the way to the FC.

**Comment by tomekste5** [comment]
I couldn't find a target for the internal module of  Boxer radio for ExpressLRS v3.0

**Comment by AndreaPro** [comment]
> If it's not too much hassle, could you guys please try ELRS 3.0? I'm not saying elrs is the issue, but we need to rule things out. As I said, the payload sent from the bf lua scripts is 4 bytes, but it could be that zero padding or something is added somewhere along the way to the FC.

I can try it and let you know

**Comment by CBockelmann** [comment]
Have the same issue with
- Meteor65, Betaflight 4.42, ELRS 3.3 LBT
- SpeedyBee F405 Mini BLS 35A 20x20 Stack, Betaflight 4.42, ELRS 3.3 LBT (BetaFPV lite receiver) and TBS Pro 32 Unify nano.

Transmitter
- FrSky QX7 ACCST with BetaFPV ELRS 1W module at ELRS 3.3 (150Hz currently due to limits of the X7)

Configuration through betaflight works fine (OSD or Configurator), just activating the LUA script even with vtx tables from betflight configurator switches the vtx channel to another band. After that nothing works and a reboot is required to reset to the old settings.

**Comment by zhornyak** [comment]
Hi 
Any update on this one?

Same issue here:
SpeedyBee F405 stack
BF 4.4.0
BF Lua 1.7.0
ELRS 2.4GHz v 3.3.0
Radiomaster TX12 Mk2 EdgeTX 2.9.1

Works on my Mobula8 without any  issues

**Comment by AndreaPro** [comment]
> > If it's not too much hassle, could you guys please try ELRS 3.0? I'm not saying elrs is the issue, but we need to rule things out. As I said, the payload sent from the bf lua scripts is 4 bytes, but it could be that zero padding or something is added somewhere along the way to the FC.
> 
> I can try it and let you know

Hello, finally I have been able to do some tests. I couldn't downgrade to ELRS 3.0 because there was no Target available for BETA FPV AIO RX. I've tried the following:

TX3.1.1 and RX3.1.1 = OK
TX3.2.1 and RX3.2.1 = OK
TX3.3.0 and RX 3.3.0 = BAD
TX 3.3.0 and RX 3.2.1 = OK

So the issue appears to happen with ELRS 3.3.0 on the RX side

**Comment by klutvott123** [comment]
@AndreaPro Thanks for testing! Maybe open an issue on the ELRS github, or ask on their discord. I'm sure they have an explanation for it.

**Comment by AndreaPro** [comment]
> @AndreaPro Thanks for testing! Maybe open an issue on the ELRS github, or ask on their discord. I'm sure they have an explanation for it.

[Done!](https://github.com/ExpressLRS/ExpressLRS/issues/2452#issue-1955368086)

**Comment by 0crap** [comment]
Same here. Latest stable v1.7.0 or nightly v20231020.10 wipes VTX tables on save.
Even reading the VTX tables goes wrong.
Reading gives:
Band 4
Channel 4
Power LV2
Pit Mode OFF
Protocol SA
Frequency 5800

While in fact my diff contains this:
# vtxtable
vtxtable bands 6
vtxtable channels 8
vtxtable band 1 BOSCAM_A A FACTORY 5865 5845 5825 5805 5785 5765 5745 5725
vtxtable band 2 BOSCAM_B B FACTORY 5733 5752 5771 5790 5809 5828 5847 5866
vtxtable band 3 BOSCAM_E E FACTORY 5705 5685 5665 5645 5885 5905 5925 5945
vtxtable band 4 FATSHARK F FACTORY 5740 5760 5780 5800 5820 5840 5860 5880
vtxtable band 5 RACEBAND R FACTORY 5658 5695 5732 5769 5806 5843 5880 5917
vtxtable band 6 LOWBAND  L CUSTOM  5362 5399 5436 5473 5510 5547 5584 5621
vtxtable powerlevels 3
vtxtable powervalues 14 20 26
vtxtable powerlabels 25 100 400

set vtx_band = 5
set vtx_channel = 3
set vtx_power = 3
set vtx_low_power_disarm = UNTIL_FIRST_ARM
set vtx_freq = 5732

**Comment by 0crap** [comment]
The <serialnumber>.lua on my SD card contains:

return {
    frequencyTable = {
        { 0, 0, 0, 0, 0, 0, 0, 0, },
        { 0, 0, 0, 0, 0, 0, 0, 0, },
        { 0, 0, 0, 0, 0, 0, 0, 0, },
        { 0, 0, 0, 0, 0, 0, 0, 0, },
        { 0, 0, 0, 0, 0, 0, 0, 0, },
        { 0, 0, 0, 0, 0, 0, 0, 0, },
    },
    frequenciesPerBand = 8,
    bandTable = { [0]="U", "1", "2", "3", "4", "5", "6", },
    powerTable = { "LV0", "LV1", "LV2", },
}

**Comment by AndreaPro** [comment]
> Same here. Latest stable v1.7.0 or nightly v20231020.10 wipes VTX tables on save. Even reading the VTX tables goes wrong. Reading gives: Band 4 Channel 4 Power LV2 Pit Mode OFF Protocol SA Frequency 5800
> 
> While in fact my diff contains this:
> 
> # vtxtable
> vtxtable bands 6 vtxtable channels 8 vtxtable band 1 BOSCAM_A A FACTORY 5865 5845 5825 5805 5785 5765 5745 5725 vtxtable band 2 BOSCAM_B B FACTORY 5733 5752 5771 5790 5809 5828 5847 5866 vtxtable band 3 BOSCAM_E E FACTORY 5705 5685 5665 5645 5885 5905 5925 5945 vtxtable band 4 FATSHARK F FACTORY 5740 5760 5780 5800 5820 5840 5860 5880 vtxtable band 5 RACEBAND R FACTORY 5658 5695 5732 5769 5806 5843 5880 5917 vtxtable band 6 LOWBAND L CUSTOM 5362 5399 5436 5473 5510 5547 5584 5621 vtxtable powerlevels 3 vtxtable powervalues 14 20 26 vtxtable powerlabels 25 100 400
> 
> set vtx_band = 5 set vtx_channel = 3 set vtx_power = 3 set vtx_low_power_disarm = UNTIL_FIRST_ARM set vtx_freq = 5732

Which drone and which ELRS version on TX and RX?

**Comment by 0crap** [comment]
@AndreaPro home made quad. ELRS 3.3.0 on both TX and RX.

**Comment by AndreaPro** [comment]
> @AndreaPro home made quad. ELRS 3.3.0 on both TX and RX.

Which FC?

**Comment by klutvott123** [comment]
it probably has something to do with the new MSP VTX code in ELRS 3.3.0

**Comment by klutvott123** [comment]
updated to 3.3.0 now. All I have to do is start the bf lua and let it download vtx tables, and then they disappear from the FC

**Comment by 0crap** [comment]
> updated to 3.3.0 now. All I have to do is start the bf lua and let it download vtx tables, and then they disappear from the FC

That is not what happens here.
Only when I hit the "save page" button from the lua script my vtxtables is missing from the diff all.
Reading alone does not ruin my vtxtables.

**Comment by klutvott123** [comment]
I just built elrs 3.3.0 without msp vtx support, and now it works fine, so there's definitely something going on in the RX code. downloading vtx tables works, and saving the vtx page works. vtx tables still there.

**Comment by bloudman648** [comment]
this problem is still there on elrs 3.3.1...will there be a fix or should we build our own fw without msp support?
and does it only ereases the vtx table? i dont trust the lua script when changing on field pids or rates...

using edgetx 2.9.1 / tx16s mark ii
iflight blitz f7 stack & blitz 1.6w vtx (nazgul v3)



btw i am new to this fpv hobby :) 
started 2 weeks ago

**Comment by klutvott123** [comment]
@bloudman648 It only affects vtx/vtx tables. Everything else should work.

There's not much we can do about this in the betaflight lua scripts. The script is working as it should and only sending the necessary bytes back when setting vtx parameters. The data is being tampered with somewhere else. 

You could report this issue here too https://github.com/ExpressLRS/ExpressLRS/issues/2452#issue-1955368086

**Comment by jonathanoeijoeng** [comment]
I have exactly same issue with this.
I'm using Radiomaster boxer using elrs 3.3.0 dan rx radiomaster rp1 same elrs version.
Everytime I download vtx table, all reset just like this case.
I did edit vtx-table.lua and copy the correct table and power but problem still remain even I have restart both tx and rx

> https://github.com/betaflight/betaflight-tx-lua-scripts#setting-up-vtx-tables

**Comment by jonathanoeijoeng** [comment]
Any luck with RX 3.3.1?

> > > If it's not too much hassle, could you guys please try ELRS 3.0? I'm not saying elrs is the issue, but we need to rule things out. As I said, the payload sent from the bf lua scripts is 4 bytes, but it could be that zero padding or something is added somewhere along the way to the FC.
> > 
> > 
> > I can try it and let you know
> 
> Hello, finally I have been able to do some tests. I couldn't downgrade to ELRS 3.0 because there was no Target available for BETA FPV AIO RX. I've tried the following:
> 
> TX3.1.1 and RX3.1.1 = OK TX3.2.1 and RX3.2.1 = OK TX3.3.0 and RX 3.3.0 = BAD TX 3.3.0 and RX 3.2.1 = OK
> 
> So the issue appears to happen with ELRS 3.3.0 on the RX side

**Comment by AndreaPro** [comment]
> Any luck with RX 3.3.1?
> 
> > > > If it's not too much hassle, could you guys please try ELRS 3.0? I'm not saying elrs is the issue, but we need to rule things out. As I said, the payload sent from the bf lua scripts is 4 bytes, but it could be that zero padding or something is added somewhere along the way to the FC.
> > > 
> > > 
> > > I can try it and let you know
> > 
> > 
> > Hello, finally I have been able to do some tests. I couldn't downgrade to ELRS 3.0 because there was no Target available for BETA FPV AIO RX. I've tried the following:
> > TX3.1.1 and RX3.1.1 = OK TX3.2.1 and RX3.2.1 = OK TX3.3.0 and RX 3.3.0 = BAD TX 3.3.0 and RX 3.2.1 = OK
> > So the issue appears to happen with ELRS 3.3.0 on the RX side

I haven't tried it yet but if no one fixed it in the Betaflight or ELRS code repos I doubt that it will be fixed by itself

**Comment by jonathanoeijoeng** [comment]
> ExpressLRS/ExpressLRS#2452 (comment)

you're right. I guess we stuck with 3.2.1 then

**Comment by LoakAudio** [comment]
Perhaps a solution on the horizon ? https://github.com/ExpressLRS/ExpressLRS/pull/2525

**Comment by haslinghuis** [comment]
Please confirm this is solved with https://github.com/ExpressLRS/ExpressLRS/pull/2525 included

**Comment by AndreaPro** [comment]
> Please confirm this is solved with [ExpressLRS/ExpressLRS#2525](https://github.com/ExpressLRS/ExpressLRS/pull/2525) included

I haven't tested it yet because I was thinking that the merged commit will become available with the next ELRS update

**Comment by haslinghuis** [comment]
I see we need to wait for 3.4 to come out.

**Comment by haslinghuis** [comment]
Ah okay, just checked, we need version 3.4 to be released for testing.

**Comment by LoakAudio** [comment]
> Ah okay, just checked, we need version 3.4 to be released for testing.

Tested today with `GIT Branch/Master` of ELRS and the penultimate `betaflight-tx-lua-scripts-nightlies`, VTX tables are loading correctly. No more issue for me.
